File: leetcode/test857/test857.go
package test857

import (
	"container/heap"
	"math"
	"sort"
)

func mincostToHireWorkers(quality []int, wage []int, k int) float64 {
	var n = len(quality)
	var arr = make([]int, 0)
	for i := 0; i < n; i++ {
		arr = append(arr, i)
	}
	sort.Slice(arr, func(i, j int) bool {
		i1, j1 := arr[i], arr[j]
		return wage[i1]*quality[j1] < wage[j1]*quality[i1]
	})
	var h = hp{make([]int, k)}
	var sum = 0
	for i := 0; i < k; i++ {
		sum += quality[arr[i]]
		h.IntSlice[i] = quality[arr[i]]
	}
	heap.Init(&h)
	var res = float64(sum) * float64(wage[arr[k-1]]) / float64(quality[arr[k-1]])
	for i := k; i < n; i++ {
		var cur = quality[arr[i]]
		if cur < h.IntSlice[0] {
			sum -= h.IntSlice[0] - cur
			h.IntSlice[0] = cur
			heap.Fix(&h, 0)
			res = math.Min(res, float64(sum)*float64(wage[arr[i]])/float64(cur))
		}
	}
	return res
}

type hp struct{ sort.IntSlice }

func (h hp) Less(i, j int) bool { return h.IntSlice[i] > h.IntSlice[j] } // 最大堆
func (hp) Push(interface{})     {}                                       // 由于没有用到，可以什么都不写
func (hp) Pop() (_ interface{}) { return }

File: leetcode/test857/Test857.java
package leetcode.test857;

import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Queue;

// q1*w2 <= w1*q2
public class Test857 {

    public double mincostToHireWorkers(int[] quality, int[] wage, int k) {
        int n = quality.length;
        int[][] arr = new int[n][2];
        for (int i = 0;i < n;i++) {
            arr[i][0] = wage[i];
            arr[i][1] = quality[i];
        }
        Arrays.sort(arr, (o1, o2) -> o1[0] * o2[1] - o1[1] * o2[0]);
        Queue<Integer> pq = new PriorityQueue<>((o1, o2) -> o2 - o1);
        double res = Double.MAX_VALUE, sum = 0;
        for (int i = 0;i < n;i++) {
            int[] cur = arr[i];
            sum -= i-k >= 0 ? pq.remove() : 0;
            sum += cur[1];
            pq.offer(cur[1]);
            double fac = (double) cur[0] / (double) cur[1];
            if (pq.size() == k) {
                res = Math.min(res, sum*fac);
            }
        }
        return res;
    }

}
